disable GSO and ECN on kernels older than version 5

marten-seemann opened the pull request:
Fixes #4396. Fixes #4178. Fixes #4446. Closes #4447.

I'm tired of the constant stream of issues with ECN and GSO. These things should just work. These are always very hard to reproduce, and mostly due to some old buggy kernel implementation that has been fixed for years, but is still out there in the wild, with no hope of ever getting updated.

New approach: disable ECN and GSO on all old (before version 5) kernels. This has some false positives, but if somebody is running such an ancient kernel, they probably don't care that much about performance anyway.

cc @heixiaoma @tobyxdd @requilence @arashpayan @ainar-g

ainar-g commented on sys_conn_helper_linux.go:
```diff
@@ -108,3 +117,40 @@ func isPermissionError(err error) bool {
 }
+
+func isECNEnabled() bool {
+	return kernelVersionMajor > 5 && !isECNDisabledUsingEnv()
```
If ECN is to follow the same logic as GSO, shouldn't this be `>=`?

ainar-g reviewed (commented):
Would you like us to try and release an edge build of AdGuard Home with quic-go from this branch?  As far as I can see, there are minor API changes as well.

marten-seemann commented:
> Would you like us to try and release an edge build of AdGuard Home with quic-go from this branch? As far as I can see, there are minor API changes as well.

@ainar-g Yes, would be great to see if this actually resolves the problem.

ainar-g commented:
@marten-seemann, I've discovered #4457 while testing, and it seems to be a blocker for us, unfortunately.  Feel free to wait for feedback from others or merge as-is.

marten-seemann commented:
> @marten-seemann, I've discovered #4457 while testing, and it seems to be a blocker for us, unfortunately. Feel free to wait for feedback from others or merge as-is.

Fixed in #4458, and rebased this PR. Can you give it another try please?

arashpayan commented:
I think this is a great idea. I'd personally take it a step further and disable it on anything older than 5.4, 5.10 or even 5.15 which are all LTS releases. 5.0 was EOL in 2019.

tobyxdd commented:
I have no problem with this, but don't we know the exact versions that added the support for these features? For example, for https://github.com/quic-go/quic-go/issues/4396 it's 4.9.

It's too shotgun to disable it on anything older than 5.0 imo

marten-seemann commented:
> I have no problem with this, but don't we know the exact versions that added the support for these features? For example, for #4396 it's 4.9.

@tobyxdd We could, but it seems like the initial implementations of both GSO and ECN contained bugs that were fixed in later (patch?) releases. Can't really fault the Linux maintainers for that, but the problem is that the update cycles are extremely long in the wild, and we still end up running on those systems years later from time to time.

> I think this is a great idea. I'd personally take it a step further and disable it on anything older than 5.4, 5.10 or even 5.15 which are all LTS releases. 5.0 was EOL in 2019.

@arashpayan Are you aware of any problems in older 5.x releases? The intent here is not to punish people for using outdated software. I just don't want to cause people problems when they run quic-go there.

marten-seemann commented on sys_conn_oob.go:
```diff
@@ -31,7 +31,7 @@ const (
 // They're both just aliases for x/net/internal/socket.Message.
 // This means we can use this struct to read from a socket that receives both IPv4 and IPv6 messages.
-var _ ipv4.Message = ipv6.Message{}
+var _ = ipv6.Message{}
```
Looks like my IDE inserted this change. This was not intended.

marten-seemann commented on sys_conn_helper_darwin.go:
```diff
@@ -33,4 +33,6 @@ func parseIPv4PktInfo(body []byte) (ip netip.Addr, ifIndex uint32, ok bool) {
-func isGSOSupported(syscall.RawConn) bool { return false }
+func isGSOEnabled(syscall.RawConn) bool { return false }
+
+func isECNEnabled() bool { return !isECNDisabledUsingEnv() }
```
I haven’t heard any complaints about macOS. Let’s wait and see?

requilence commented:
The GSO control message was introduced in kernel version 4.18. According to my observations, many Android devices still use the 4.19 kernel. Additionally, Google [still supports](https://source.android.com/docs/core/architecture/kernel/android-common#compatibility-matrix) upgrading devices with the 4.19 kernel to the latest Android 14. However, I haven't found any public data on the distribution of Android kernels in the wild.

So, at the end, I think disabling GSO on kernels older than 5.0 in the quic-go library is reasonable and should improve the stability, as the Linux kernel repo contains numerous fixes for GSO between 4.18 and 5.0

bt90 commented:
> According to my observations, many Android devices still use the 4.19 kernel.

Depending on the manufacturer, the switch to the 5.x kernel only happened on devices that **launched** with Android 11 or 12. So if the device is older than ~2 years, chances are it's still running the 4.x kernel.

![image](https://github.com/quic-go/quic-go/assets/1926175/4e3ca0c8-4bc4-4488-9042-db37f54e78c1)

Source: https://source.android.com/docs/core/architecture/kernel/android-common#feature-and-launch-kernels

That said, blocking GSO/ECN on anything pre 5.x is still the right call. quic-go can't and shouldn't try to unfuck the Android ecosystem.
